parser-typechecker/src/Unison/Codebase/Branch.hs

```diff
@@ -211,80 +220,163 @@ branch0 terms types children edits =
+        go :: Seq (DeepChildAcc m) -> Set Path -> DeepState m (Set Path)
+        go Seq.Empty acc = pure acc
+        go (e@(reversePrefix, _, b0) Seq.:<| work) acc = do
+          let paths :: Set Path
+              paths =
+                if isEmpty0 b0
+                  then Set.empty
+                  else (Set.singleton . Path . Seq.fromList . reverse) reversePrefix
+          children <- deepChildrenHelper e
+          go (work <> children) (paths <> acc)
 
 -- | Derive the 'deepEdits' field of a branch.
-deriveDeepEdits :: Branch0 m -> Branch0 m
+deriveDeepEdits :: forall m. Branch0 m -> Branch0 m
 deriveDeepEdits branch =
-  branch {deepEdits = makeDeepEdits (_edits branch) (nonEmptyChildren branch)}
+  branch {deepEdits = makeDeepEdits branch}
   where
-    makeDeepEdits :: Map NameSegment (EditHash, m Patch) -> Map NameSegment (Branch m) -> Map Name EditHash
-    makeDeepEdits edits children =
-      Map.mapKeysMonotonic Name.fromSegment (Map.map fst edits) <> ifoldMap go children
+    makeDeepEdits :: Branch0 m -> Map Name EditHash
+    makeDeepEdits branch = State.evalState (go (Seq.singleton ([], 0, branch)) mempty) Set.empty
       where
-        go :: NameSegment -> Branch m -> Map Name EditHash
-        go n b =
-          Map.mapKeys (Name.cons n) (deepEdits $ head b)
+        go :: (Seq (DeepChildAcc m)) -> Map Name EditHash -> DeepState m (Map Name EditHash)
+        go Seq.Empty acc = pure acc
+        go (e@(reversePrefix, _, b0) Seq.:<| work) acc = do
+          let edits :: Map Name EditHash
+              edits =
+                Map.mapKeysMonotonic
+                  (Name.fromReverseSegments . (NonEmpty.:| reversePrefix))
+                  (fst <$> _edits b0)
+          children <- deepChildrenHelper e
+          go (work <> children) (edits <> acc)
+
+type DeepState m = State (Set (NamespaceHash m))
+
+type DeepChildAcc m = ([NameSegment], Int, Branch0 m)
+
+-- | Helper for knowing whether to descend into a child branch or not.
+-- Accepts child namespaces with previously unseen hashes, and any nested under fewer than 2 `lib` segments.
+deepChildrenHelper :: forall m. DeepChildAcc m -> DeepState m (Seq (DeepChildAcc m))
+deepChildrenHelper (reversePrefix, libDepth, b0) = do
+  let libDepth' = case reversePrefix of
+        "lib" : _ -> libDepth + 1
+        _other -> libDepth
+  let go :: (NameSegment, Branch m) -> DeepState m (Seq (DeepChildAcc m))
+      go (ns, b) = do
+        let h = namespaceHash b
+        result <- do
+          let isDirectDependency = libDepth' < 2
+          isUnseenTransitiveDep <- State.gets (Set.notMember h)
+          pure
+            if isDirectDependency || isUnseenTransitiveDep
+              then Seq.singleton (ns : reversePrefix, libDepth', head b)
+              else Seq.empty
+        State.modify' (Set.insert h)
+        pure result
+  Monoid.foldMapM go (Map.toList (nonEmptyChildren b0))
+
+-- | Flattens a Metadata.Star into a 4-tuple.
```

Comment:
I see this is almost `Metadata.starToR4`, except that returns this list in `R4`'d form. Maybe this belongs in that module anyway though, as `Metadata.starToR4List` or something?